proposals/nnnn-structured-concurrency.md

```diff
@@ -255,6 +257,148 @@ If the scope of an `async let` exits with a thrown error, the child task corresp
+  }
+}
+```
+
+A nursery _guarantees_ that it will `await` for all tasks that were added to it before it returns.
+
+This waiting can be performed either: 
+- by the code within the nursery itself, or
+- by transparently nursery itself when returning from it.
+
+In the `chopVegetables()` example we not only added vegetable chopping tasks to the nursery, but also collected the chopped up results. See below for simplified reminder of the general pattern:
+
+```swift
+func chopVegetables(rawVeggies: [Vegetable]) async throws -> [ChoppedVegetable] {
+  Task.withNursery(resultType: ChoppedVegetable.self) { nursery in    
+    var choppedVeggies: [ChoppedVegetable] = []
+    choppedVeggies.reserveCapacity(veggies.count)
+        
+    // add all chopping tasks and process them concurrently
+    for v in rawVeggies {
+      await try nursery.add { // await the successful adding of the task 
+        await v.chopped() // await the processing result of task
+      }
+    }
+
+    while let choppedVeggie = await try nursery.next() { 
+      choppedVeggies.append(choppedVeggie)
+    }
+    
+    return choppedVeggies
+  }
+}
+```
+
+#### Nurseries: Throwing and cancellation
+
+Worth pointing out here is that adding a task to a nursery could fail because the nursery could have been cancelled when we were about to add more tasks to it. To visualize this, let us consider the following example:
+
+Tasks in a nursery by default handle thrown errors using like the musketeers would, that is: "*One for All, and All for One!*" In other words, if a single task throws an error, which escapes into the nursery, all other tasks will be cancelled and the nursery will re-throw this error.
+
+To visualize this, let us consider chopping vegetables again. One type veggetable that can be quite tricky to chop up is onions, they can make you cry if you don't watch out. If we attempt to chop up those vegetables, the onion will throw an error into the nursery, causing all other tasks to be cancelled automatically:
+
+```swift
+func chopOnionsAndCarrots(rawVeggies: [Vegetable]) async throws -> [Vegetable] {
+  await try Task.withNursery { nursery in // (3) will re-throw the onion chopping error
+    // kick off asynchronous vegetable chopping:
+    for v in rawVeggies {
+      await try nursery.add { 
+        await try v.chopped() // (1) throws
+      }
+    }
+    
+    // collect chopped up results:
+    while let choppedVeggie = await try nursery.next() { // (2) will throw for the onion
+      choppedVeggies.append(choppedVeggie)
+    }
+  }
+}
+```
+
+Let us break up the `chopOnionsAndCarrots()` function into multiple steps to fully understand its semantics:
+
+- first w add vegetable chopping tasks to the nursery
+- the chopping of the various vegetables beings asynchronously,
+- eventually an onion will be chopped and `throw`
+
+
+So far we did not yet discuss the cancellation of nurseries. A nursery can be cancelled form the outside, for example like this: 
```

Comment:
So we currently _don't_ have an API proposed to make a nursery and then launch stuff with it...

```swift
// "kind of" exists but we didn't really design it completely IMHO
let nursery = Nursery()

await try nursery.add {  // awaits only the "add"
 await task() 
}
await try nursery.add {  // awaits only the "add"
 await task() 
}

sleep(200)

nursery.cancel() // maybe some other thread
```

but also one could:

```swift
let handle = Task.runDetached { 
Task.withNursery { ... }
}

handle.cancel()
```

which is what I was going to show here for now.

Reply:
I don't think we'd have the first, I don't want arbitrary code doing stuff with nurseries.

Reply:
Okey got it, I wonder if that will really be doable but going to assume so for now then -- will only doc the parent task getting cancelled then 👍

Reply:
I documented the latter: https://github.com/DougGregor/swift-evolution/pull/13/commits/19625c4e65d3d3a66ea09663b796077ae7a61079 -- comment on the commit with thoughts...